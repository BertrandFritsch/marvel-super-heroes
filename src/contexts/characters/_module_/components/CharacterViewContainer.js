// @flow

import type { State } from '../flowTypes';
import type { Dispatch } from 'redux';

import { connect } from 'react-redux';
import actions from '../actions';
import CharacterView from './CharacterView';
import CharacterTiles from './CharacterTilesContainer';
import CharacterDetails from './CharacterDetailsContainer';

const tilesViewProps = {
  title: 'Liste des super héros',
  component: CharacterTiles
};

const detailsViewProps = {
  title: 'Fiche d\'identité',
  component: CharacterDetails
};

/**
 * Select the view to render:
 *  - the character details if a character is selected
 *  - the character tiles otherwise
 */
export default connect(
  (state: State) => state.characters.get('selectedItem') ? detailsViewProps : tilesViewProps,
  (dispatch: Dispatch<*>) => ({
    onNavigationRequested: state => dispatch(actions.onNavigationRequested(state))
  })
)(CharacterView);
