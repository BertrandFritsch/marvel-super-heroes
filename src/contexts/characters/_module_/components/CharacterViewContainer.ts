import { connect } from 'react-redux';
import ActionTypes, { Dispatch } from '../actionTypes';
import { CharacterState, State } from '../reducers';
import CharacterDetails from './CharacterDetailsContainer';
import CharacterTiles from './CharacterTilesContainer';
import CharacterView from './CharacterView';

const tilesViewProps = {
  title: 'Liste des super héros',
  component: CharacterTiles
};

const detailsViewProps = {
  title: 'Fiche d\'identité',
  component: CharacterDetails
};

/**
 * Select the view to render:
 *  - the character details if a character is selected
 *  - the character tiles otherwise
 */
export default connect(
  (state: CharacterState) => state.characters.selectedItem ? detailsViewProps : tilesViewProps,
  (dispatch: Dispatch<{ characters: State }>) => ({
    onNavigationRequested: (state: string | null): void => dispatch({
      type: ActionTypes.NAVIGATION_REQUESTED,
      payload: state
    })
  }),
  (stateProps, dispatchProps) => ({
    ...stateProps,
    ...dispatchProps
  })
)(CharacterView);
