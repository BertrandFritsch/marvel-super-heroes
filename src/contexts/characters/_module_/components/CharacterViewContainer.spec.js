import React from 'react';
import { Provider } from 'react-redux';
import { Map, List } from 'immutable';
import CharacterViewContainer from './CharacterViewContainer';

describe('CharacterViewContainer', () => {
  let store;
  let state;

  beforeEach(() => {
    state = {
      characters: Map({
        items: List(),
        selectedItem: null
      })
    };
    store = ({
      default: () => {
      },
      subscribe: () => {
      },
      dispatch: jest.fn(),
      getState: () => state
    });
  });

  it('should render correctly', () => {
    const wrapper = mount(
      <Provider store={ store }>
        <CharacterViewContainer />
      </Provider>
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('should render the tiles view', () => {
    state.characters = state.characters.set('items', List([
      {
        id: 'id1',
        name: 'name1',
        thumbnail: 'http://.../standard_xlarge.jpg',
        thumbnail_details: 'http://.../portrait_incredible.jpg',
        urls: [
          {
            type: 'details',
            url: 'http://...'
          }
        ]
      }
    ]));

    const wrapper = mount(
      <Provider store={ store }>
        <CharacterViewContainer />
      </Provider>
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('should render the tiles view while loading', () => {
    state.characters = state.characters.set('loadingItems', 'true');

    const wrapper = mount(
      <Provider store={ store }>
        <CharacterViewContainer />
      </Provider>
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('should render the details view', () => {
    state.characters = state.characters
      .set('items', List([
        {
          id: 'id1',
          name: 'name1',
          thumbnail: 'http://.../standard_xlarge.jpg',
          thumbnail_details: 'http://.../portrait_incredible.jpg',
          urls: [
            {
              type: 'details',
              url: 'http://...'
            }
          ]
        }
      ]))
      .set('selectedItem', {
        id: 'id1',
        name: 'name1',
        thumbnail: 'http://.../standard_xlarge.jpg',
        thumbnail_details: 'http://.../portrait_incredible.jpg',
        urls: [
          {
            type: 'details',
            url: 'http://...'
          }
        ],
        comics: {
          items: [
            {
              name: 'comics1'
            }
          ]
        },
        series: {
          items: [
            {
              name: 'series1'
            }
          ]
        }
      });

    const wrapper = mount(
      <Provider store={ store }>
        <CharacterViewContainer />
      </Provider>
    );

    expect(wrapper).toMatchSnapshot();
  });
});
