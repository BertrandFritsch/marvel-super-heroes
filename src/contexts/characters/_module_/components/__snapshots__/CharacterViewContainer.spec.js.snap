// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CharacterViewContainer should render correctly 1`] = `
<Provider
  store={
    Object {
      "default": [Function],
      "dispatch": [Function],
      "getState": [Function],
      "subscribe": [Function],
    }
  }
>
  <Connect(CharacterView)>
    <CharacterView
      component={[Function]}
      onNavigationRequested={[Function]}
      title="Liste des super héros"
    >
      <div
        className="main-container"
      >
        <nav>
          <span
            className="marvel-logo"
          >
            Liste des super héros
          </span>
        </nav>
        <section
          className="character-section"
        >
          <Connect(CharacterTiles)>
            <CharacterTiles
              items={Array []}
              message={null}
              onTileDetailsRequested={[Function]}
              onTilesViewLoaded={[Function]}
            >
              <_class
                columnWidth={250}
                duration={800}
                gutterHeight={10}
                gutterWidth={10}
                itemHeight={280}
              >
                <_class
                  angleUnit="deg"
                  columnWidth={250}
                  columns={4}
                  component="div"
                  duration={800}
                  easing="cubic-bezier(0.215, 0.61, 0.355, 1)"
                  enter={[Function]}
                  entered={[Function]}
                  exit={[Function]}
                  gutterHeight={10}
                  gutterWidth={10}
                  itemHeight={280}
                  layout={[Function]}
                  lengthUnit="px"
                >
                  <ReactTransitionGroup
                    childFactory={[Function]}
                    component="div"
                    style={
                      Object {
                        "height": "-10px",
                        "position": "relative",
                        "width": "1030px",
                      }
                    }
                  >
                    <div
                      style={
                        Object {
                          "height": "-10px",
                          "position": "relative",
                          "width": "1030px",
                        }
                      }
                    />
                  </ReactTransitionGroup>
                </_class>
              </_class>
            </CharacterTiles>
          </Connect(CharacterTiles)>
        </section>
      </div>
    </CharacterView>
  </Connect(CharacterView)>
</Provider>
`;

exports[`CharacterViewContainer should render the details view 1`] = `
<Provider
  store={
    Object {
      "default": [Function],
      "dispatch": [Function],
      "getState": [Function],
      "subscribe": [Function],
    }
  }
>
  <Connect(CharacterView)>
    <CharacterView
      component={[Function]}
      onNavigationRequested={[Function]}
      title="Fiche d'identité"
    >
      <div
        className="main-container"
      >
        <nav>
          <span
            className="marvel-logo"
          >
            Fiche d'identité
          </span>
        </nav>
        <section
          className="character-section"
        >
          <Connect(CharacterDetails)>
            <CharacterDetails
              dispatch={[Function]}
              item={
                Object {
                  "comics": Object {
                    "items": Array [
                      Object {
                        "name": "comics1",
                      },
                    ],
                  },
                  "id": "id1",
                  "name": "name1",
                  "series": Object {
                    "items": Array [
                      Object {
                        "name": "series1",
                      },
                    ],
                  },
                  "thumbnail": "http://.../standard_xlarge.jpg",
                  "thumbnail_details": "http://.../portrait_incredible.jpg",
                  "urls": Array [
                    Object {
                      "type": "details",
                      "url": "http://...",
                    },
                  ],
                }
              }
            >
              <div
                className="character-details"
              >
                <img
                  src="http://.../portrait_incredible.jpg"
                />
                <div
                  className="character-details-desc"
                >
                  <span
                    className="character-details-name"
                  >
                    name1
                  </span>
                  <p
                    className="character-details-description"
                  />
                  <CharacterDetailsCollection
                    items={
                      Array [
                        Object {
                          "name": "comics1",
                        },
                      ]
                    }
                    name="Comics"
                  >
                    <div
                      className="character-collection"
                    >
                      <span
                        className="character-collection-name"
                      >
                        Comics
                      </span>
                      <div
                        className="character-collection-items"
                      >
                        <span
                          className="character-collection-item"
                        >
                          comics1
                        </span>
                      </div>
                    </div>
                  </CharacterDetailsCollection>
                  <CharacterDetailsCollection
                    items={
                      Array [
                        Object {
                          "name": "series1",
                        },
                      ]
                    }
                    name="Series"
                  >
                    <div
                      className="character-collection"
                    >
                      <span
                        className="character-collection-name"
                      >
                        Series
                      </span>
                      <div
                        className="character-collection-items"
                      >
                        <span
                          className="character-collection-item"
                        >
                          series1
                        </span>
                      </div>
                    </div>
                  </CharacterDetailsCollection>
                </div>
              </div>
            </CharacterDetails>
          </Connect(CharacterDetails)>
        </section>
      </div>
    </CharacterView>
  </Connect(CharacterView)>
</Provider>
`;

exports[`CharacterViewContainer should render the tiles view 1`] = `
<Provider
  store={
    Object {
      "default": [Function],
      "dispatch": [Function],
      "getState": [Function],
      "subscribe": [Function],
    }
  }
>
  <Connect(CharacterView)>
    <CharacterView
      component={[Function]}
      onNavigationRequested={[Function]}
      title="Liste des super héros"
    >
      <div
        className="main-container"
      >
        <nav>
          <span
            className="marvel-logo"
          >
            Liste des super héros
          </span>
        </nav>
        <section
          className="character-section"
        >
          <Connect(CharacterTiles)>
            <CharacterTiles
              items={
                Array [
                  Object {
                    "id": "id1",
                    "name": "name1",
                    "thumbnail": "http://.../standard_xlarge.jpg",
                    "thumbnail_details": "http://.../portrait_incredible.jpg",
                    "urls": Array [
                      Object {
                        "type": "details",
                        "url": "http://...",
                      },
                    ],
                  },
                ]
              }
              message={null}
              onTileDetailsRequested={[Function]}
              onTilesViewLoaded={[Function]}
            >
              <_class
                columnWidth={250}
                duration={800}
                gutterHeight={10}
                gutterWidth={10}
                itemHeight={280}
              >
                <_class
                  angleUnit="deg"
                  columnWidth={250}
                  columns={4}
                  component="div"
                  duration={800}
                  easing="cubic-bezier(0.215, 0.61, 0.355, 1)"
                  enter={[Function]}
                  entered={[Function]}
                  exit={[Function]}
                  gutterHeight={10}
                  gutterWidth={10}
                  itemHeight={280}
                  layout={[Function]}
                  lengthUnit="px"
                >
                  <ReactTransitionGroup
                    childFactory={[Function]}
                    component="div"
                    style={
                      Object {
                        "height": "280px",
                        "position": "relative",
                        "width": "1030px",
                      }
                    }
                  >
                    <div
                      style={
                        Object {
                          "height": "280px",
                          "position": "relative",
                          "width": "1030px",
                        }
                      }
                    >
                      <_class
                        angleUnit="deg"
                        columnWidth={250}
                        columns={4}
                        component="div"
                        duration={800}
                        easing="cubic-bezier(0.215, 0.61, 0.355, 1)"
                        enter={[Function]}
                        entered={[Function]}
                        exit={[Function]}
                        gridProps={
                          Object {
                            "angleUnit": "deg",
                            "children": Array [
                              <div>
                                <CharacterTile
                                      item={
                                            Object {
                                                  "id": "id1",
                                                  "name": "name1",
                                                  "thumbnail": "http://.../standard_xlarge.jpg",
                                                  "thumbnail_details": "http://.../portrait_incredible.jpg",
                                                  "urls": Array [
                                                    Object {
                                                      "type": "details",
                                                      "url": "http://...",
                                                    },
                                                  ],
                                                }
                                      }
                                      onTileDetailsRequested={[Function]}
                                />
                          </div>,
                            ],
                            "columnWidth": 250,
                            "columns": 4,
                            "component": "div",
                            "duration": 800,
                            "easing": "cubic-bezier(0.215, 0.61, 0.355, 1)",
                            "enter": [Function],
                            "entered": [Function],
                            "exit": [Function],
                            "gutterHeight": 10,
                            "gutterWidth": 10,
                            "itemHeight": 280,
                            "layout": [Function],
                            "lengthUnit": "px",
                          }
                        }
                        gridState={
                          Object {
                            "gridHeight": 280,
                            "gridWidth": 1030,
                            "positions": Array [
                              Array [
                                0,
                                0,
                              ],
                            ],
                          }
                        }
                        gutterHeight={10}
                        gutterWidth={10}
                        itemHeight={280}
                        layout={[Function]}
                        lengthUnit="px"
                        position={
                          Array [
                            0,
                            0,
                          ]
                        }
                        transition="opacity 800ms cubic-bezier(0.215, 0.61, 0.355, 1), transform 800ms cubic-bezier(0.215, 0.61, 0.355, 1)"
                      >
                        <div
                          style={
                            Object {
                              "WebkitTransform": "translateX(0px) translateY(0px) scale(1)",
                              "WebkitTransition": "opacity 800ms cubic-bezier(0.215, 0.61, 0.355, 1), transform 800ms cubic-bezier(0.215, 0.61, 0.355, 1)",
                              "left": 0,
                              "opacity": 1,
                              "position": "absolute",
                              "top": 0,
                              "transform": "translateX(0px) translateY(0px) scale(1)",
                              "transition": "opacity 800ms cubic-bezier(0.215, 0.61, 0.355, 1), transform 800ms cubic-bezier(0.215, 0.61, 0.355, 1)",
                              "zIndex": 2,
                            }
                          }
                        >
                          <CharacterTile
                            item={
                              Object {
                                "id": "id1",
                                "name": "name1",
                                "thumbnail": "http://.../standard_xlarge.jpg",
                                "thumbnail_details": "http://.../portrait_incredible.jpg",
                                "urls": Array [
                                  Object {
                                    "type": "details",
                                    "url": "http://...",
                                  },
                                ],
                              }
                            }
                            onTileDetailsRequested={[Function]}
                          >
                            <div
                              className="character-tile"
                            >
                              <a
                                href="#"
                                onClick={[Function]}
                              >
                                <img
                                  src="http://.../standard_xlarge.jpg"
                                />
                              </a>
                              <span
                                className="character-tile-name"
                              >
                                name1
                              </span>
                              <div
                                className="character-tile-links"
                              >
                                <a
                                  href="http://..."
                                >
                                  details
                                </a>
                              </div>
                            </div>
                          </CharacterTile>
                        </div>
                      </_class>
                    </div>
                  </ReactTransitionGroup>
                </_class>
              </_class>
            </CharacterTiles>
          </Connect(CharacterTiles)>
        </section>
      </div>
    </CharacterView>
  </Connect(CharacterView)>
</Provider>
`;

exports[`CharacterViewContainer should render the tiles view while loading 1`] = `
<Provider
  store={
    Object {
      "default": [Function],
      "dispatch": [Function],
      "getState": [Function],
      "subscribe": [Function],
    }
  }
>
  <Connect(CharacterView)>
    <CharacterView
      component={[Function]}
      onNavigationRequested={[Function]}
      title="Liste des super héros"
    >
      <div
        className="main-container"
      >
        <nav>
          <span
            className="marvel-logo"
          >
            Liste des super héros
          </span>
        </nav>
        <section
          className="character-section"
        >
          <Connect(CharacterTiles)>
            <CharacterTiles
              items={Array []}
              message="Chargement des héros..."
              onTileDetailsRequested={[Function]}
              onTilesViewLoaded={[Function]}
            >
              <CharacterMessage
                message="Chargement des héros..."
              >
                <p
                  className="character-message"
                >
                  Chargement des héros...
                </p>
              </CharacterMessage>
            </CharacterTiles>
          </Connect(CharacterTiles)>
        </section>
      </div>
    </CharacterView>
  </Connect(CharacterView)>
</Provider>
`;
